[DB-590] Live subscriptions: Re-authorize when a stream's metadata changes

shaan1337 opened the pull request:
Changed: Re-authorize stream access in live subscriptions when stream metadata changes

Note that:
i)  some live events can still go through before the AccessDenied message is sent to the subscription.
ii) user group changes are not taken into consideration. this is OK as with claims-based authorization, existing access is maintained until the claims expire.

Note: .NET client tests pass against this branch.

hayley-jean commented on src/EventStore.Core/Services/SubscriptionsService.cs:
```diff
@@ -335,6 +350,69 @@ public class SubscriptionsService<TStreamId> :
 		}
 
+		private void ProcessStreamMetadataChanges(string eventStreamId) {
+			if (eventStreamId == AllStreamsSubscriptionId)
```
It might be worth moving these conditions to ProcessEventCommitted so that this method isn't entered for every event

shaan1337 replied:
i've moved `ProcessStreamMetadataChanges()` to `Handle(StorageMessage.EventCommitted message)` / `Handle(StorageMessage.InMemoryEventCommitted message)`. this gets rid of the unnecessary `AllStreamsSubscriptionId` empty string check

for the other checks (`SystemStreams.IsMetastream()` / `_subscriptionTopics.TryGetValue()`), i've kept them in the ProcessStreamMetadataChanges(). even though it's on a hot path, i don't think that it should cause any performance penalty as entering/leaving the method is negligible and the method will also be inlined automatically by the JIT compiler if it's called frequently.

shaan1337 replied:
it also raises the question as to whether or not we should process stream metadata changes for ephemeral streams. i've tested adding stream metadata to the $mem-node-state ephemeral stream and it's working fine.

timothycoleman replied:
the memstream thing is fine by me

(moving the conditions was my idea. i was thinking the target method might be on the large side for inlining and and the checks will almost always return false it seemed worth it,   but now that we're calling it from two places, and you're right its probably not going to make much practical diference, lets leave it as is)

timothycoleman commented:
i think we can use 'Changed' rather than 'Fixed' in the description (probaly dont worry about the branch name though)

timothycoleman commented:
i pushed some small changes, after this is rebased im happy to approve

YoEight commented:
How difficult would it be to have an unit test for this?

shaan1337 commented:
@YoEight i've added some tests for it later in https://github.com/EventStore/EventStore/pull/4108 (`subscribe * then revoke access`). I just squashed & rebased this PR to master, then rebased #4108 and launched the new tests - they're passing.
